// bolt/dist/modern/utils/options.js
'use strict';

exports.__esModule = true;
exports.string = string;
exports.boolean = boolean;
exports.number = number;
exports.toDependencyTypes = toDependencyTypes;
exports.toSpawnOpts = toSpawnOpts;
exports.toFilterOpts = toFilterOpts;
exports.toDependency = toDependency;
exports.toYarnInit = toYarnInit;

var _constants = require('../constants');

function string(val, name) {
  if (typeof val !== 'undefined' && typeof val !== 'string') {
    throw new Error(`Flag "${name}" must be string`);
  } else {
    return val;
  }
}
function boolean(val, name) {
  if (typeof val !== 'undefined' && typeof val !== 'boolean') {
    throw new Error(`Flag "${name}" must be boolean`);
  } else {
    return val;
  }
}

function number(val, name) {
  if (typeof val !== 'undefined' && typeof val !== 'number') {
    throw new Error(`Flag "${name}" must be number`);
  } else {
    return val;
  }
}

function toDependencyTypes(val, name) {
  if (typeof val !== 'string') {
    throw new Error(`Flag "${name}" must be a string`);
  }
  let types = val.split(',');
  let invalidTypes = types.filter(t => _constants.DEPENDENCY_TYPES.indexOf(t) === -1);
  if (invalidTypes.length > 0) {
    throw new Error(`Flag "${name}" must be a comma separated list of valid dependency types. Received invalid types: "${invalidTypes.join(',')}"`);
  }
  return types;
}

function toSpawnOpts(flags) {
  let spawnOpts = {};

  if ((flags.parallelNodes || flags.parallel) && flags.serial) {
    throw new Error('Commands cannot be run both serially and in parallel');
  }

  if (flags.parallelNodes) spawnOpts.orderMode = 'parallel-nodes';
  if (flags.parallel) spawnOpts.orderMode = 'parallel';
  if (flags.serial) spawnOpts.orderMode = 'serial';
  if (typeof flags.bail !== 'undefined') spawnOpts.bail = boolean(flags.bail, 'bail');
  if (flags.excludeFromGraph) spawnOpts.excludeFromGraph = toDependencyTypes(flags.excludeFromGraph, 'excludeFromGraph');
  // TODO:
  // if (flags.concurrency) spawnOpts.maxConcurrent = number(flags.concurrency, 'concurrency');

  return spawnOpts;
}

function toFilterOpts(flags) {
  let filterOpts = {};

  if (flags.only) filterOpts.only = string(flags.only, 'only');
  if (flags.onlyFs) filterOpts.onlyFs = string(flags.onlyFs, 'onlyFs');
  if (flags.ignore) filterOpts.ignore = string(flags.ignore, 'ignore');
  if (flags.ignoreFs) filterOpts.ignoreFs = string(flags.ignoreFs, 'ignoreFs');

  return filterOpts;
}

/**
 * Takes a string in one of the following forms:
 *  "pkgName", "pkgName@version", "@scope/pkgName", "@scope/pkgName@version"
 * and returns an object with the package name and version (if passed)
 */
function toDependency(dependencyString) {
  let [name, ...rest] = dependencyString.split('@').filter(part => part !== '');
  let version = rest.join('@');
  if (name.includes('/')) {
    name = '@' + name;
  }
  return version.length > 0 ? { name, version } : { name };
}

function toYarnInit(flags) {
  let flagOpts = {};

  if (flags.yes || flags.y) flagOpts.yes = true;
  if (flags.private || flags.p) flagOpts.private = true;

  return flagOpts;
}
