// bolt/dist/legacy/GlobalOptions.js
'use strict';

exports.__esModule = true;
exports.globalOptions = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _options = require('./utils/options');

var options = _interopRequireWildcard(_options);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GlobalOptionsStore = function () {
  function GlobalOptionsStore() {
    (0, _classCallCheck3.default)(this, GlobalOptionsStore);
    this.store = {
      disableCmdPrefix: undefined
    };
  }

  GlobalOptionsStore.prototype.set = function set(key, value) {
    this.store[key] = value;
  };

  GlobalOptionsStore.prototype.get = function get(key) {
    return this.store[key];
  };

  GlobalOptionsStore.prototype.getAll = function getAll() {
    return this.store;
  };

  GlobalOptionsStore.prototype.setFromFlags = function setFromFlags(flags) {
    if (flags.prefix === false) {
      this.set('disableCmdPrefix', true);
    }
  };

  return GlobalOptionsStore;
}(); /**
      * Singleton that manages global bolt options.
      */

var globalOptions = exports.globalOptions = new GlobalOptionsStore();
